src/extra/gfx/drawable_utils.hpp

```diff
@@ -2,56 +2,57 @@
 #define C7AA640C_1748_42BC_8772_0B8F4E136CEA
 
 #include "gfx/error_utils.hpp"
+#include "gfx/shader/types.hpp"
+#include "shards.h"
 #include "shards_types.hpp"
 #include "shards_utils.hpp"
 #include <ops_internal.hpp>
 #include <foundation.hpp>
 #include <gfx/material.hpp>
 #include <gfx/params.hpp>
+#include <gfx/texture.hpp>
 #include <magic_enum.hpp>
+#include <optional>
 #include <shards.hpp>
 #include <spdlog/spdlog.h>
+#include <variant>
 
 namespace gfx {
-inline void varToTexture(const SHVar &var, TexturePtr &outVariant) {
-  switch (var.valueType) {
-  case SHType::Object: {
-    shards::Type type = shards::Type::Object(var.payload.objectVendorId, var.payload.objectTypeId);
-    if (type == Types::Texture) {
-      auto ptr = reinterpret_cast<TexturePtr *>(var.payload.objectValue);
-      assert(ptr);
-      outVariant = *ptr;
-    } else {
-      throw formatException("Expected texture object");
-    }
-  } break;
-  default:
-    throw formatException("Value type {} can not be converted to TextureVariant", magic_enum::enum_name(var.valueType));
-  };
+inline TexturePtr varToTexture(const SHVar &var) {
+  if (var.payload.objectTypeId == Types::TextureCubeTypeId) {
+    return *varAsObjectChecked<TexturePtr>(var, Types::TextureCube);
+  } else if (var.payload.objectTypeId == Types::TextureTypeId) {
+    return *varAsObjectChecked<TexturePtr>(var, Types::Texture);
+  } else {
+    SHInstanceData data{};
+    auto varType = shards::deriveTypeInfo(var, data);
+    DEFER({ shards::freeDerivedInfo(varType); });
```

Comment:
Why defer?
Ah error message

Reply:
Yeah I want to display the type in the error